pages/studentAssessments/studentAssessments.sql

```diff
@@ -120,12 +120,17 @@ WITH
       FULL JOIN assessments AS a ON (gc.assessment_id = a.id)
       JOIN course_instances AS ci ON (ci.id = a.course_instance_id)
       JOIN assessment_sets AS aset ON (aset.id = a.assessment_set_id)
-      LEFT JOIN assessment_instances AS ai ON (
-        ai.assessment_id = a.id
-        AND (
-          ai.user_id = $user_id
-          OR ai.group_id = gu.group_id
-        )
+      -- We join assessment_instances twice, once on user_id and once on
+      -- group_id. We used to have these combined with AND (ai.user_id =
+      -- $user_id OR ai.group_id = gu.group_id) but this was triggering a bad
+      -- query plan for some course instances.
+      LEFT JOIN assessment_instances AS ai1 ON (
+        ai1.assessment_id = a.id
+        AND ai1.user_id = $user_id
+      )
+      LEFT JOIN assessment_instances AS ai2 ON (
+        ai2.assessment_id = a.id
+        AND ai2.group_id = gu.group_id
       )
```

Comment:
```suggestion
      LEFT JOIN (
        SELECT * FROM assessment_instances AS ai1 WHERE
        ai1.assessment_id = a.id
        AND ai1.user_id = $user_id
        UNION ALL
        SELECT * FROM assessment_instances AS ai2 WHERE
        ai2.assessment_id = a.id
        AND ai2.group_id = gu.group_id
      ) AS ai ON (TRUE)
```

Not sure how Postgres would react to this, but I believe this may make the query a bit clearer.

Reply:
https://www.cybertec-postgresql.com/en/avoid-or-for-better-performance/

Reply:
Good thinking! I switched to this in 7bdee2d54007fc661ffbe98b90bb2792aebbc34e. I tested this and it still avoids seq scans. This modified query should behave identically to the old code, avoiding the edge case I identified above: https://github.com/PrairieLearn/PrairieLearn/pull/7381#discussion_r1153512721

Reply:
@jonatanschroeder that's an excellent blog post, thanks!